### DISCUSSION TITLE
How to override the Optimization level for the Release and Debug build.

### DISCUSSION THREAD

**Discussion by AndreaCatania** [jrouwe/JoltPhysics#1076] [Q&A] [answered]
I would like to be able to turn the optimization level to `Od`, for release builds, because of what mentioned [here](https://github.com/jrouwe/JoltPhysics/pull/1075):
> Unreal Engine has two type of compilation targets DebugGame and Development. When compiling in DebugGame, despite the name, the produced binary is not a target Debug but it's still target Release but without optimizations.

I've read it's not really possible to control the optimization level using the configuration flags.
I also tried by specifying it on the flags using the following code, but didn't work:

```
			cmakeOptions += " -DCMAKE_C_FLAGS_RELEASE=\"/GS /Od /Ob0 /RTC1\" ";
			cmakeOptions += " -DCMAKE_CXX_FLAGS_RELEASE=\"/GS /Od /Ob0 /RTC1\" ";
			cmakeOptions += " -DCMAKE_C_FLAGS=\"/GS /Od /Ob0 /RTC1\" ";
			cmakeOptions += " -DCMAKE_CXX_FLAGS=\"/GS /Od /Ob0 /RTC1\" ";
```

I wonder what's the be way to add this feature in Jolt. Do you think an IF [condition here](https://github.com/jrouwe/JoltPhysics/blob/master/Build/CMakeLists.txt#L129) is ok?

Something like this:
```CMake
if (DEBUG_RELEASE)
    set(CMAKE_CXX_FLAGS_RELEASE "/GS /Od /Ob0 /RTC1")
else()
    set(CMAKE_CXX_FLAGS_RELEASE "/GS- /Gy /O2 /Oi /Ot")
endif()
```

**Comment by jrouwe** [comment]
If Unreal uses a config called DebugGame then I'd expect you to need to override `CMAKE_CXX_FLAGS_DEBUGGAME` instead of `CMAKE_CXX_FLAGS_RELEASE`.

**Reply by AndreaCatania** [reply]
Unreal doesn't use CMake to build the engine and the game. It uses the Unreal Build Tool, written in C#. Jolt is build by unreal, using CMake and then the produced binary is linked.

The cmakeOptions, you see above, is used to configure the Jolt build.

The idea I have in mind is to add an additional config flag to jolt's CMakeList, to control the optimization level, so the produced binary can be debugged. (I've already tried changing the opt level manually by modifying the CMakList, and it works. Changing optimization level allows me to debug the jolt code from unreal. I'm just not sure what implementation you prefer, to allow me configuring it from the Unreal Build Tool).

**Reply by jrouwe** [reply, marked as answer]
In that case you can use `OVERRIDE_CXX_FLAGS` to skip setting the cxx flags from Jolt's CMakeLists.txt so that you can set it yourself (note this used to work for clang only, I just submitted a fix that makes this work for MSVC too).

**Reply by AndreaCatania** [reply]
That's perfect, I'll try it tomorrow. Thanks!!

**Reply by AndreaCatania** [reply]
@jrouwe This works perfectly, thanks!

Here the code I've used to enable it:

```
cmakeOptions += " -DOVERRIDE_CXX_FLAGS=OFF ";
cmakeOptions += " -DCMAKE_CXX_FLAGS_RELEASE=\"/GS /Od /Ob0 /RTC1\" ";
```

I would have expected to ENABLE the override using `OVERRIDE_CXX_FLAGS=ON` and not to disable it, since from the game project prospective I'm overriding what Jolt is providing. Other than that it works really well!

**Reply by jrouwe** [reply]
> I would have expected to ENABLE the override using OVERRIDE_CXX_FLAGS=ON

That's unfortunately a breaking change, which is something I try to avoid as much as possible.

**Reply by AndreaCatania** [reply]
Yeah, that's fine. No need to change it, it works really well already.
